File: packages/shared-types/src/conversation.ts
export type MessageRole = "user" | "assistant";

export interface Source {
  document_id: string;
  document_name: string;
  page_number?: number;
  chunk_text: string;
  relevance_score: number;
}

export interface TokenUsage {
  prompt: number;
  completion: number;
  total: number;
}

export interface Message {
  id: string;
  conversation_id: string;
  role: MessageRole;
  content: string;
  sources: Source[];
  model_used?: string;
  tokens_used?: TokenUsage;
  created_at: string;
}

export interface Conversation {
  id: string;
  title?: string;
  created_at: string;
  updated_at: string;
  messages?: Message[];
}

export interface AskRequest {
  question: string;
  options?: {
    top_k?: number;
    score_threshold?: number;
    document_filter?: string[];
  };
}

export interface AskResponse {
  answer: string;
  sources: Source[];
  model_used: string;
  tokens_used: TokenUsage;
}

export interface ConversationListResponse {
  conversations: Conversation[];
  total: number;
}

File: apps/api/src/app/schemas/conversation.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class SourceResponse(BaseModel):
    document_id: str
    document_name: str
    page_number: int | None = None
    chunk_text: str
    relevance_score: float


class TokenUsage(BaseModel):
    prompt: int = 0
    completion: int = 0
    total: int = 0


class MessageResponse(BaseModel):
    id: UUID
    conversation_id: UUID
    role: str
    content: str
    sources: list[SourceResponse] = []
    model_used: str | None = None
    tokens_used: TokenUsage | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class ConversationCreate(BaseModel):
    title: str | None = None


class ConversationResponse(BaseModel):
    id: UUID
    title: str | None = None
    created_at: datetime
    updated_at: datetime
    messages: list[MessageResponse] = []

    model_config = {"from_attributes": True}


class ConversationListResponse(BaseModel):
    conversations: list[ConversationResponse]
    total: int


class AskRequest(BaseModel):
    question: str = Field(..., min_length=1, max_length=5000)
    options: dict = Field(default_factory=dict)


class AskResponse(BaseModel):
    answer: str
    sources: list[SourceResponse]
    model_used: str
    tokens_used: TokenUsage
